services/placement/selector/mirrored.go

```diff
@@ -0,0 +1,465 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package selector
+
+import (
+	"container/heap"
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+
+	"github.com/m3db/m3x/log"
+
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/placement"
+)
+
+var (
+	errNoValidMirrorInstance = errors.New("no valid instance for mirror placement in the candidate list")
+)
+
+type mirroredFilter struct {
+	opts   services.PlacementOptions
+	logger xlog.Logger
+}
+
+func newMirroredSelector(opts services.PlacementOptions) placement.InstanceSelector {
+	return &mirroredFilter{
+		opts:   opts,
+		logger: opts.InstrumentOptions().Logger(),
+	}
+}
+
+// SelectInitialInstances tries to make as many groups as possible from
+// the candidate instances to make the initial placement.
+func (f *mirroredFilter) SelectInitialInstances(
+	candidates []services.PlacementInstance,
+	rf int,
+) ([]services.PlacementInstance, error) {
+	candidates, err := getValidCandidates(placement.NewPlacement(), candidates, f.opts)
+	if err != nil {
+		return nil, err
+	}
+
+	weightToHostMap, err := groupHostsByWeight(candidates)
+	if err != nil {
+		return nil, err
+	}
+
+	var groups = make([][]services.PlacementInstance, 0, len(candidates))
+	for _, hosts := range weightToHostMap {
+		groupedHosts, ungrouped := groupHostsWithRackCheck(hosts, rf)
+		if len(ungrouped) != 0 {
+			for _, host := range ungrouped {
```

Comment:
Okay yeah I think mostly likely we'll pull hosts from a pool, so I'm fine w/ leaving it as is.